src/compiler.ml

```diff
@@ -3899,17 +4024,18 @@ end = struct
     F.Map.union (fun k _ _ -> error ("Duplicate type abbreviation for " ^ F.show k)) m1 m2
 
   let extend1 flags
-    (state, { Assembled.hash; symbols; prolog_program; indexing; modes = om; kinds = ok; types = ot; type_abbrevs = ota; chr = ochr; toplevel_macros = otlm; total_type_checking_time })
-            { version; base_hash; checked_code = { CheckedFlat.toplevel_macros; kinds; types; types_indexing; type_abbrevs; modes; clauses; chr; builtins}; precomputed_kinds; precomputed_type_abbrevs; precomputed_types; type_checking_time } =
+    (state, { Assembled.hash; symbols; prolog_program; indexing; modes = om; kinds = ok; functional_preds = ofp; types = ot; type_abbrevs = ota; chr = ochr; toplevel_macros = otlm; total_type_checking_time })
+            { version; base_hash; checked_code = { CheckedFlat.toplevel_macros; kinds; types; types_indexing; type_abbrevs; modes; functional_preds; clauses; chr; builtins}; precomputed_kinds; precomputed_type_abbrevs; precomputed_types; type_checking_time } =
     let symbols, prolog_program, indexing = update_indexing state symbols prolog_program modes types_indexing indexing in
-    let kinds, type_abbrevs, types =
-    if hash = base_hash then
-      precomputed_kinds, precomputed_type_abbrevs, precomputed_types
-    else
-      let kinds = Flatten.merge_kinds ok kinds in
-      let type_abbrevs = merge_type_abbrevs ota type_abbrevs in
-      let types = Flatten.merge_type_assignments ot types in
-      kinds, type_abbrevs, types
+    let kinds, type_abbrevs, types, functional_preds =
+      if hash = base_hash then
+        precomputed_kinds, precomputed_type_abbrevs, precomputed_types, functional_preds
```

Comment:
The idea is that type checkig requires to merge some stuff. the precomputed_stuff is already merged.
I have some use cases in which the base used for type checking is not the base used for linking (a bit unsafe at the moment).
If the base is the same for both steps, the already merged stuff can be used. If not, one has to merge since the base changed and the precomputed values are wrong.

If it is also the case for functional_preds, follow the name schema/field positions a the other stuff.
If not, don't computed functional_preds in this if-then-else.